core/cmd/local_client.go

```diff
@@ -128,20 +130,54 @@ func (cli *Client) runNode(c *clipkg.Context) error {
 		lggr.Warn("Ethereum is disabled. Chainlink will only run services that can operate without an ethereum connection")
 	}
 
+	// DB conn is opened before any services have started,
+	// and closed after all services have stopped.
 	db, err := openDB(cli.Config, lggr)
 	if err != nil {
 		return cli.errorOut(errors.Wrap(err, "opening db"))
 	}
 	defer lggr.ErrorIfClosing(db, "db")
 
-	sig := shutdown.NewSignal()
-	cleanup, err := applicationLockDB(cli.Config, db, lggr, sig)
-	defer cleanup()
+	// rootCtx will be cancelled when SIGINT|SIGTERM is received.
+	rootCtx, cancelRootCtx := context.WithCancel(context.Background())
+
+	// cleanExit is used to skip "fail fast" routine
+	cleanExit := make(chan struct{})
+	var shutdownStartTime time.Time
+	defer func() {
+		close(cleanExit)
+		log.Printf("Graceful shutdown time: %s", time.Since(shutdownStartTime))
+	}()
+
+	go shutdown.HandleShutdown(func() {
+		shutdownStartTime = time.Now()
+		cancelRootCtx()
```

Comment:
Does this ensure that locks are only released after the app has finished shutting down?

Reply:
Discussed in a private call.